mmdet3d/ops/paconv/paconv.py

```diff
@@ -0,0 +1,350 @@
+                 kernel_input: str = 'neighbor',
+                 scorenet_cfg: dict = dict(
+                     mlp_channels=[8, 16, 16],
+                     score_norm='softmax',
+                     temp_factor=1.0,
+                     last_bn=False)):
+        super(PAConv, self).__init__()
+
+        # determine weight kernel size according to used features
+        if kernel_input == 'identity':
+            # only use grouped_features
+            self.kernel_mul = 1
+        elif kernel_input == 'neighbor':
+            # concat of (grouped_features - center_features, grouped_features)
+            self.kernel_mul = 2
+        else:
+            raise NotImplementedError(
+                f'unsupported kernel_input {kernel_input}')
+        self.kernel_input = kernel_input
+
+        # determine mlp channels in ScoreNet according to used xyz features
+        if scorenet_input == 'identity':
+            # only use relative position (grouped_xyz - center_xyz)
+            self.scorenet_in_channels = 3
+        elif scorenet_input == 'neighbor':
+            # (grouped_xyz - center_xyz, grouped_xyz)
+            self.scorenet_in_channels = 6
+        elif scorenet_input == 'ed7':
+            # (center_xyz, grouped_xyz - center_xyz, Euclidian distance)
+            self.scorenet_in_channels = 7
+        else:
+            raise NotImplementedError(
+                f'unsupported scorenet_input {scorenet_input}')
+        self.scorenet_input = scorenet_input
+
+        # construct weight kernels in weight bank
+        # self.weight_bank is of shape [C, num_kernels * out_c]
+        # where C can be in_c or (2 * in_c)
+        if weight_bank_init == 'kaiming':
+            weight_init = nn.init.kaiming_normal_
+        elif weight_bank_init == 'xavier':
+            weight_init = nn.init.xavier_normal_
+        else:
+            raise NotImplementedError(
+                f'unsupported weight bank init method {weight_bank_init}')
+
+        self.m = num_kernels
+        weight_bank = weight_init(
+            torch.empty(self.m, in_channels * self.kernel_mul, out_channels))
+        weight_bank = weight_bank.permute(1, 0, 2).reshape(
+            in_channels * self.kernel_mul, self.m * out_channels).contiguous()
+        self.weight_bank = nn.Parameter(weight_bank, requires_grad=True)
+
+        # construct ScoreNet
+        scorenet_cfg_ = copy.deepcopy(scorenet_cfg)
+        scorenet_cfg_['mlp_channels'].insert(0, self.scorenet_in_channels)
+        scorenet_cfg_['mlp_channels'].append(self.m)
+        self.scorenet = ScoreNet(**scorenet_cfg_)
+
+        self.bn = build_norm_layer(norm_cfg, out_channels)[1] if \
+            norm_cfg is not None else None
+        self.activate = build_activation_layer(act_cfg) if \
+            act_cfg is not None else None
+
+    def forward(self, points_xyz: torch.Tensor,
+                features: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
```

Comment:
Actually we do not need such specific types for these input params because we have explanations in the docstring. Some previous codes keep these (due to some legacy issues) but we can just simplify them in our new codes.
Just write as forward(self, points_xyz, features) to be more concise.

Reply:
Fix PAConv and PAConvCUDA's forward func.